src/modlunky2/ui/trackers/runstate.py

```diff
@@ -100,15 +116,130 @@ def update_no_gold(self, run_recap_flags):
         if not bool(run_recap_flags & RunRecapFlags.NO_GOLD):
             self.run_label.discard(Label.NO_GOLD)
 
-    def update_no_tp(self, player_item_types):
-        if not self.no_tp:
+    def update_no_tp(
+        self,
+        game_state: State,
+        player: Player,
+        player_item_set: Set[EntityType],
+        prev_player_item_set: Set[EntityType],
+    ):
+        prev_next_uid = self.prev_next_uid
+        self.prev_next_uid = game_state.next_entity_uid
+        # At the start of a level, all entities (including floors, treasure, etc.) spawn.
+        # Also, the player doesn't gain control for ~600ms anyway
+        if self.level_started:
+            return
+
+        # This is an optimization to skip scanning when the player couldn't have teleported
+        if not self.could_tp(player, player_item_set, prev_player_item_set):
             return
 
-        for item_type in player_item_types:
-            if item_type in TELEPORT_ENTITIES:
-                self.no_tp = False
+        found_shadows: List[LightEmitter] = []
+        for shadow_uid in range(prev_next_uid, game_state.next_entity_uid):
+            shadow_poly = game_state.instance_id_to_pointer.get(shadow_uid)
+            if shadow_poly is None or not shadow_poly.present():
+                continue
+            shadow = shadow_poly.value
```

Comment:
Nit: These `shadow_` variable names are slightly confusing to me because these are any and all entities at this point, not just shadows.

Reply:
made these `entity_foo`